### PR TITLE
Make dropout condition on training.

### PR DESCRIPTION
**Pull request by ailzhang**

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -2179,6 +2179,28 @@ def test_dropout(self):
...
+    def test_dropout_requires_grad(self):
+        class MyModule(torch.nn.Module):
+            def __init__(self, M):
+                super(MyModule, self).__init__()
+
+                self.dropout = torch.nn.Dropout(0.5)
+                self.linear = torch.nn.Linear(M, M)
+
+            def forward(self, input):
+                input = F.dropout(input, 0.5, self.training)
```
Comment (SsnL) [inline test/test_jit.py]: did you mean to use `self.dropout` here?

**Review by ssnl — COMMENTED** [review]
wow nice catch! pretty big bug I'd say though.....

File: torch/csrc/jit/symbolic_script.cpp [hunk 2]
```diff
@@ -1170,12 +1170,11 @@ const std::vector<std::string> functions = {
             if use_cuda:
```
Comment (bddppq) [inline torch/csrc/jit/symbolic_script.cpp]: I think we can move this into the “if train” case as well?

File: torch/csrc/jit/symbolic_script.cpp [hunk 3]
```diff
@@ -1170,12 +1170,11 @@ const std::vector<std::string> functions = {
...
+            elif train:
                 mask = torch.empty_like(input)
                 mask.bernoulli_(p1m)
                 res = mask * input / p1m
-
-            if not train:
+            else:
                 p1m = 1.
                 res = input
                 mask = torch.ones_like(input)
```
Comment (bddppq) [inline torch/csrc/jit/symbolic_script.cpp]: Could we avoid computing mask in non-train case?

**Comment by bddppq** [comment]
Thanks for fixing

File: test/test_jit.py [hunk 4]
```diff
@@ -2179,6 +2179,28 @@ def test_dropout(self):
             m = self.createFunctionFromGraph(g)
             self.assertEqual(outputs, m(*inputs))
 
+    def test_dropout_requires_grad(self):
+        class MyModule(torch.nn.Module):
+            def __init__(self, M):
+                super(MyModule, self).__init__()
+
+                self.dropout = torch.nn.Dropout(0.5)
```
Comment (bddppq) [inline test/test_jit.py]: I think we should check both the module version and the functional version.

File: test/test_jit.py [hunk 5]
```diff
@@ -2179,6 +2179,34 @@ def test_dropout(self):
...
+
+        M = 1000
+        scripted_module = torch.jit.script(MyModule(M))
+        scripted_func = torch.jit.script(func)
+
+        X = torch.randn(M, M, requires_grad=True)
+        FileCheck().check("aten::bernoulli_").run(scripted_module.graph_for(X))
+        FileCheck().check("aten::bernoulli_").run(scripted_func.graph_for(X))
+
+        X = torch.randn(M, M, requires_grad=False)
```
Comment (SsnL) [inline test/test_jit.py]: The condition here is a bit weird. It should not depend on whether `x` is requiring grad, but on whether `m` is in training mode, right?

Comment (zdevito) [reply test/test_jit.py]: The torchscript executor only bothers to split out symbolic gradients when the inputs require_grad, which is probably why these flags are getting set here.

**Review by zdevito — APPROVED** [review]

File: test/test_jit.py (hunk 5 above)
Comment (SsnL) [reply test/test_jit.py]: My question was more about why there is not a .eval() here to switch the scripted module to eval mode. Unless .graph_for auto selected train / eval mode based on input requires_grad which sounds wrong to me.

Comment (ailzhang) [reply test/test_jit.py]: @SsnL @zdevito @bddppq Thanks for the comments! To make the testcase a bit more clear, I've added a comment here to explain the behavior. https://github.com/pytorch/pytorch/pull/29436/files#diff-78aab5b4217aeaed920efab0c5f5f5c8R2202
Please let me know if that looks good! Thanks!

Comment (SsnL) [reply test/test_jit.py]: Thank you so much!!!
